### PR TITLE
Documentation dataset format

### PR DESCRIPTION
**Pull request by qgallouedec**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### New section in the doc

- Added new section about dataset format in the doc, directly review in the online doc: https://moon-ci-docs.huggingface.co/docs/trl/pr_2020/en/dataset_formats

### New data utils

- Added new data utils: `maybe_apply_chat_template`, `maybe_extract_prompt`, `maybe_unpair_preference_dataset`, `apply_chat_template`, `extract_prompt` and `unpair_preference_dataset`.
- Added tests for all of them

### Update dataset example files

As explained [in this section](https://moon-ci-docs.huggingface.co/docs/trl/pr_2020/en/dataset_formats#using-any-dataset-with-trl-preprocessing-and-conversion) we provide script for converting dataset to TRL style.
The goal is, for every dataset in [`trl-lib`](https://huggingface.co/trl-lib) to have its corresponding script in `examples/datasets`

- Updated all files in `examples/datasets` to closely match the documented format.

### Start to update some example scripts

We have to update all scripts to make sure they comply with the new style. 

- Updated DPO script

 I prefer dedicating further PR to update them all

- Updated the `SIMPLE_QUERY_CHAT_TEMPLATE` to include a prompt generation

### CODE REVIEW

**Review by kashif — APPROVED** [review]

**Review by kashif — APPROVED** [review]

**Review by kashif — APPROVED** [review]

File: docs/source/dataset_formats.mdx [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+| Trainer                 | Expected dataset format      |
+| ----------------------- | ---------------------------- |
+| [`BCOTrainer`]          | Unpaired preference          |
+| [`CPOTrainer`]          | Preference (explicit prompt) |
+| [`DPOTrainer`]          | Preference (explicit prompt) |
+| [`IterativeSFTTrainer`] | Unpaired preference          |
+| [`KTOTrainer`]          | Unpaired preference          |
+| [`OnlineDPOTrainer`]    | Prompt-only                  |
+| [`ORPOTrainer`]         | Preference (explicit prompt) |
+| [`PPOv2Trainer`]        | Tokenized language modeling  |
```
Comment (lewtun) [inline docs/source/dataset_formats.mdx:202]: For some reason the link to this trainer doesn't render

File: docs/source/dataset_formats.mdx [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+| [`RewardTrainer`]       | Preference (implicit prompt) |
+| [`SFTTrainer`]          | Language modeling            |
+| [`XPOTrainer`]          | Prompt-only                  |
```
Comment (lewtun) [inline docs/source/dataset_formats.mdx:205]: Same here

File: docs/source/dataset_formats.mdx [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+</Tip>
+
+<Tip warning={true}>
+
+It's important to note that chat templates are model-specific. For example, if you use the chat template from [meta-llama/Meta-Llama-3.1-8B-Instruct](https://huggingface.co/meta-llama/Meta-Llama-3.1-8B-Instruct) with the above example, you get a different output:
+
+```python
+apply_chat_template(example, AutoTokenizer.from_pretrained("meta-llama/Meta-Llama-3.1-8B-Instruct"))
+# Output:
+# {'prompt': '<|im_start|>system\nYou are a helpful assistant.<|im_end|>\n<|im_start|>user\nWhat color is the sky?<|im_end|>\n<|im_start|>assistant\n',
```
Comment (lewtun) [inline docs/source/dataset_formats.mdx:277]: This looks like the wrong chat template - the llama one is different

File: docs/source/dataset_formats.mdx [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+
+Many datasets come in formats tailored to specific tasks, which might not be directly compatible with TRL. To use such datasets with TRL, you may need to preprocess and convert them into the required format.
+
+To make this easier, we provide a set of [example scripts](https://github.com/huggingface/trl/tree/main/examples/datasets) that cover common dataset conversions.
+
+### Example: UltraFeedback dataset
+
+Let’s take the [UltraFeedback dataset](https://huggingface.co/datasets/openbmb/UltraFeedback) as an example. Here's a preview of the dataset:
+
+<iframe
```
Comment (lewtun) [inline docs/source/dataset_formats.mdx:295]: Awesome use of the dataset viewer!!

File: docs/source/dataset_formats.mdx [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+For more information on how to work with conversational datasets, refer to the [Working with conversational datasets in TRL](#working-with-conversational-datasets-in-trl) section.
+
+</Tip>
+
+## Working with conversational datasets in TRL
+
+Conversational datasets are increasingly common, especially for training chat models. However, TRL trainers don't support conversational datasets in their raw format. These datasets must first be converted into a standard format. 
+Fortunately, TRL offers tools to easily handle this conversion, which are detailed below.
+
+### Converting a conversational dataset into a standard dataset
```
Comment (lewtun) [inline docs/source/dataset_formats.mdx:219]: I'm a bit confused about this mapping since I thought the `SFTTrainer` allows users to provide a dataset with a `messages` column and no preprocessing / conversion is needed. 

Overall, I think we'd like to have a situation where users can have datasets formatted in the conversational style and the chat templating step is taken care of automatically. I realise we don't have consistency in the lib right now, so I'm open to alternative ideas!

File: examples/scripts/dpo.py [hunk 6]
```diff
@@ -128,7 +131,7 @@
     if tokenizer.pad_token is None:
         tokenizer.pad_token = tokenizer.eos_token
     if tokenizer.chat_template is None:
-        tokenizer.chat_template = "{% for message in messages %}{{message['role'] + ': ' + message['content'] + '\n\n'}}{% endfor %}{{ eos_token }}"
+        tokenizer.chat_template = SIMPLE_QUERY_CHAT_TEMPLATE
```
Comment (lewtun) [inline examples/scripts/dpo.py:134]: Nice catch!

File: tests/test_data_utils.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+
+    @parameterized.expand(itertools.product(non_conversational_examples))
+    def test_non_conversational(self, example):
+        self.assertFalse(is_conversational(example))
+
+
+class ApplyChatTemplateTester(unittest.TestCase):
+    tokenizers = [
+        "Qwen/Qwen2-7B-Instruct",
+        # "meta-llama/Meta-Llama-3.1-8B-Instruct", # gated repository. TODO: find a way to test this
```
Comment (lewtun) [inline tests/test_data_utils.py:89]: You can use one of the tiny random models from the transformers team, e.g. https://huggingface.co/hf-internal-testing/tiny-random-LlamaForCausalLM

File: trl/trainer/utils.py [hunk 8]
```diff
@@ -881,7 +881,8 @@ def print_rich_table(df: pd.DataFrame) -> Table:
 SIMPLE_SFT_CHAT_TEMPLATE = "{% for message in messages %}{{' ' + message['content']}}{% endfor %}{{ eos_token }}"
 # SIMPLE_SFT_CHAT_TEMPLATE simply ends things with an EOS token, this helps the SFT model learn to end the completions with EOS tokens
 
-SIMPLE_QUERY_CHAT_TEMPLATE = "{% for message in messages %}{{' ' + message['content']}}{% endfor %}"
+SIMPLE_QUERY_CHAT_TEMPLATE = "{% for message in messages %}{{message['role'] + ': ' + message['content'] + '\n\n'}}{% endfor %}{% if add_generation_prompt %}{{ 'assistant:' }}{% endif %}"
```
Comment (lewtun) [inline trl/trainer/utils.py:884]: Would a good default be the one from Anthropic, which has alternating `Human:` and `Assistant:` roles? https://huggingface.co/datasets/Anthropic/hh-rlhf

**Comment by qgallouedec** [comment]
> Everything LGTM, with one main question about what is meant by "standard dataset format".

Yes... "standard" is in fact "non-conversational". But it's weird to define something by something it's not.
It's not easy to find relevant, accurate and easy-to-understand names. If anyone have any suggestions for changes, I'd be happy to use them.

> What I'm wondering in particular is whether we expect users to preformat their datasets for each trainer or whether we accept some formats like `messages` as column that are automatically formatted in our scripts.

Currently we expect users to preformat their datasets. But I think supporting everything in the trainers can make sense:

```python
from datasets import Dataset

standard_dataset = Dataset.from_dict(
    {
        "prompt": ["The sky is", "The sun is"],
        "completion": [" blue.", " in the sky."],
    }
)
AnyTrlTrainer(..., train_dataset=standard_dataset)  # ok


conversational_dataset = Dataset.from_dict({
    "prompt": [
        [{"role": "user", "content": "What color is the sky?"}], [{"role": "user", "content": "Where is the sun?"}],
    ],
    "completion": [
        [{"role": "assistant", "content": "It is blue."}], [{"role": "assistant", "content": "In the sky."}],
    ],
})
AnyTrlTrainer(..., train_dataset=conversational_dataset)  # currently not ok, but we can support it in the future.
```

File: docs/source/dataset_formats.mdx [hunk 9]
```diff
@@ -0,0 +1,709 @@
...
+example = {"text": "The sky is blue."}
+# Preference
+example = {"chosen": "The sky is blue.", "rejected": "The sky is green."}
+```
+
+### Conversational dataset format
+
+Conversational datasets are used for tasks involving dialogues or chat interactions between users and assistants. Unlike standard datasets format, these contain sequences of messages where each message has a `role` (e.g., `"user"` or `"assistant"`) and `content` (the message text).
+
+```python
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply docs/source/dataset_formats.mdx:95]: yes https://github.com/huggingface/trl/pull/2020#discussion_r1754153844

File: docs/source/dataset_formats.mdx (hunk 5 above)
Comment (qgallouedec) [reply docs/source/dataset_formats.mdx:219]: It's annoying because `SFTTrainer` is a particular case: it supports 
- tokenized dataset
- standard dataset (ie. non-conversational)
- and conversational dataset

But it's the only trainer to support all of these.  Another reason to extend and standardise this to all trainers.

I will mention this particular case in the doc though.

File: trl/trainer/utils.py (hunk 8 above)
Comment (qgallouedec) [reply trl/trainer/utils.py:884]: Something like this:?

```suggestion
SIMPLE_QUERY_CHAT_TEMPLATE = "{% for message in messages %}{% if message['role'] == 'user' %}Human: {% elif message['role'] == 'assistant' %}Assistant: {% endif %}{{ message['content'] }} \n\n{% endfor %}{% if add_generation_prompt %}Assistant:{% endif %}"
```

```python
>>> from transformers import AutoTokenizer
>>> tokenizer = AutoTokenizer.from_pretrained("gpt2")
>>> tokenizer.chat_template = "{% for message in messages %}{% if message['role'] == 'user' %}Human: {% elif message['role'] == 'assistant' %}Assistant: {% endif %}{{ message['content'] }} \n\n{% endfor %}{% if add_generation_prompt %}Assistant:{% endif %}"
>>> messages = [
...     {"role": "user", "content": "Hello, how are you?"},
...     {"role": "assistant", "content": "I'm doing great. How can I help you today?"},
...     {"role": "user", "content": "I'd like to show off how chat templating works!"},
... ]
>>> print(tokenizer.apply_chat_template(messages, tokenize=False))
Human: Hello, how are you? 

Assistant: I'm doing great. How can I help you today? 

Human: I'd like to show off how chat templating works! 
```

**Review by kashif — APPROVED** [review]

File: docs/source/dataset_formats.mdx [hunk 10]
```diff
@@ -0,0 +1,709 @@
...
+    <td>
+      <pre><code>{"prompt": [{"role": "user", "content": "What color is the sky?"}],
+ "completion": [{"role": "assistant", "content": "It is green."}],
+ "label": False}</code></pre>
+    </td>
+  </tr>
+</table>
+
+
+### Standard dataset format
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply docs/source/dataset_formats.mdx:82]: Clarified in [c2b6574](https://github.com/huggingface/trl/pull/2020/commits/c2b657456e510aee9b4327c6e6a72b4f5fc00c58)

File: trl/trainer/utils.py (hunk 8 above)
Comment (qgallouedec) [reply trl/trainer/utils.py:884]: Let's just capitalise the role, so that it keep things simple (we don't really need the user to be converted to human))

```suggestion
SIMPLE_QUERY_CHAT_TEMPLATE = "{% for message in messages %}{{message['role'].capitalize() + ': ' + message['content'] + '\n\n'}}{% endfor %}{% if add_generation_prompt %}{{ 'Assistant:' }}{% endif %}"
```

File: tests/test_data_utils.py (hunk 7 above)
Comment (qgallouedec) [reply tests/test_data_utils.py]: Adressed in [040d4cf](https://github.com/huggingface/trl/pull/2020/commits/040d4cf2eac76298898389cca4c7db94b3a8bb1b)

**Comment by lewtun** [comment]
> Yes... "standard" is in fact "non-conversational".

Ah, that makes sense - thanks for the clarification. One thing I'm wondering is whether there's any real advantage to support standard vs conversational, since technically all standard datasets can be converted to conversational by wrapping the prompt, completion etc in the messages format. 

Given that most models are chat models, one approach would be to support conversational datasets natively in the trainers, but allow users to also provide a preprocessed / tokenized dataset if they wish more flexibility. 

This also is related to #1646
